enclave/src/polkadex_gateway.rs

```diff
@@ -0,0 +1,137 @@
+    // Authenticate
+    authenticate_user(main_account.clone()  , proxy_acc)?;
+    // TODO: Send cancel order to Openfinex API
+    // Mutate Balances
+    match polkadex_orderbook_storage::remove_order(&order_uuid)? {
+        Some(cancelled_order) => {
+            match cancelled_order.order_type {
+                OrderType::LIMIT => {
+                    if let Some(price) = cancelled_order.price {
+                        match cancelled_order.side {
+                            OrderSide::BID => {
+                                // TODO: Is it safe to use saturating_mul here?
+                                let amount = price.saturating_mul(cancelled_order.quantity);
+                                polkadex_balance_storage::unreserve_balance(main_account.clone(), cancelled_order.market_id.quote, amount)?;
+                            }
+                            OrderSide::ASK => {
+                                // TODO: Is it safe to use saturating_mul here?
+                                polkadex_balance_storage::unreserve_balance(main_account.clone(), cancelled_order.market_id.base, cancelled_order.quantity)?;
+                            }
+                        }
+                    } else {
+                        error!("Unable to find price for limit order");
+                        return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+                    }
+                }
+                OrderType::MARKET => {}
+                OrderType::FillOrKill | OrderType::PostOnly => {
+                    error!("OrderType is not implemented");
+                    return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+                }
+            }
+        }
+        None => {
+            // TODO: Is this error handling correct? Since,
+            // TODO: we already sent the request to cancel to Openfinex but
+            // TODO: order was not found in Enclave's Orderbook Storage
+            error!("Unable to find order for given order_uuid");
+            return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+        }
+    }
+    Ok(())
+}
+
+
+pub fn authenticate_user(main_acc: AccountId, proxy_acc: Option<AccountId>) -> SgxResult<()> {
+    // Authentication
+    match proxy_acc {
+        Some(proxy) => {
+            if !polkadex::check_if_proxy_registered(main_acc, proxy)? {
+                error!("Proxy Account is not registered for given Main Account");
+                return Err(sgx_status_t::SGX_ERROR_UNEXPECTED); // TODO: It would be better if we can return a error response for RPC to handle
+            }
+        }
+        None => {
+            if !polkadex::check_if_main_account_registered(main_acc)? {
+                error!("Main Account is not registered");
+                return Err(sgx_status_t::SGX_ERROR_UNEXPECTED); // TODO: It would be better if we can return a error response for RPC to handle
```

Comment:
how can we return a sensible response back to RPC when it calls this function

Reply:
Resolved

Reply:
Return a normal Result, not Sgx. And then define the error you want.. we just have to send it back with a JSON String, but with encoding pretty much everything (from struct to enum to string) can be sent back to the client